### PR TITLE
introduce rw-pass-cell

### PR DESCRIPTION
**Pull request by pepyakin**
See the documentation. Will be used in the following commits.

### CODE REVIEW

**Comment by pepyakin** [comment]
* **#73** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/73?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#72** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/72?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#71** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/71?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pepyakin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/rw_pass_cell.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            _write_pass: write_pass,
+        }
+    }
+
+    /// Prevents any funny business by ensuring the domain of the pass and the domain of the cell
+    /// match.
+    fn check_domain(&self, domain: &Arc<Shared>) {
+        let domain_ptr = Arc::as_ptr(domain);
+        let provenance_ptr = Weak::as_ptr(&self.provenance);
+        assert!(std::ptr::eq(domain_ptr, provenance_ptr), "Domain mismatch");
```
Comment (rphmeier) [inline nomt/src/rw_pass_cell.rs:155]: awesome. I was looking for exactly this

File: nomt/src/rw_pass_cell.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    'pass: 'a,
+{
+    pub fn get(&self) -> &T {
+        // SAFETY: The existence of the guard ensures that there is only one mutable reference
+        //         to the inner value. The returned reference cannot outlive the guard.
+        unsafe { &*self.inner.inner.get() }
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        // SAFETY: The existence of the guard ensures that there is only one mutable reference
```
Comment (rphmeier) [inline nomt/src/rw_pass_cell.rs:205]: worth noting: The existence of the _mutable reference to_ the guard

**Review by rphmeier — APPROVED** [review]
I had the seed of an idea like this as of last night, but this looks absolutely fantastic and even has a domain ptr_eq check. Nicely done.

File: nomt/src/rw_pass_cell.rs (hunk 1 above)
Comment (rphmeier) [reply nomt/src/rw_pass_cell.rs:155]: #77

File: nomt/src/rw_pass_cell.rs (hunk 2 above)
Comment (rphmeier) [reply nomt/src/rw_pass_cell.rs:205]: #77

**Comment by rphmeier** [comment]
### Merge activity

* **Apr 5, 8:06 PM EDT**: @rphmeier started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thrumdev/nomt/71).

**Pull request merged**